jax/_src/lax/control_flow/conditionals.py

```diff
@@ -433,6 +435,78 @@ def _cond_partial_eval(trace, *tracers, branches, linear):
+def _cond_partial_eval_custom(saveable, unks_in, inst_in, eqn):
+  index_uk, *ops_uk = unks_in
+  assert not index_uk  # only possible with old-style remat
+  branches = eqn.params['branches']
+
+  # First, compute output unknowns (unks_out), where an output of the cond is
+  # unknown if it would be unknown on any of the branches.
+  unks_out: List[bool] = [False] * len(eqn.outvars)
+  for jaxpr in branches:
+    _, _, unks_out_, _, _ = pe.partial_eval_jaxpr_custom(
+        jaxpr.jaxpr, in_unknowns=ops_uk, in_inst=[True] * len(ops_uk),
+        ensure_out_unknowns=False, ensure_out_inst=True, saveable=saveable)
+    unks_out = map(operator.or_, unks_out, unks_out_)
+
+  # Next, use the computed output unknowns to build a known jaxpr and a staged
+  # jaxpr for each branch.
+  branches_known : List[core.ClosedJaxpr] = []
+  branches_staged: List[core.ClosedJaxpr] = []
+  branch_res_avals: List[core.AbstractValue] = []
+  for jaxpr in branches:
+    jaxpr_known, jaxpr_staged, _, inst_out, num_res = \
+        pe.partial_eval_jaxpr_custom(
+            jaxpr.jaxpr, in_unknowns=ops_uk, in_inst=[True] * len(ops_uk),
+            ensure_out_unknowns=unks_out, ensure_out_inst=True,
+            saveable=saveable)
+    branches_known.append( core.ClosedJaxpr(jaxpr_known,  jaxpr.consts))
+    branches_staged.append(core.ClosedJaxpr(jaxpr_staged, jaxpr.consts))
+    branch_res_avals.append(branches_staged[-1].in_avals[:num_res])
+
+  # Residuals may differ across branches, so we merge them, then use the merged
+  # residuals to join the outputs of all branches to the same type.
+  all_res_avals, res_avals_per_branch = _merge_branch_residuals(branch_res_avals)
+  num_res = len(all_res_avals)
+  num_known_outs = len(unks_out) - sum(unks_out)
+  branches_known = _join_cond_outputs(
+      branches_known, all_res_avals, res_avals_per_branch, num_known_outs)
+  branches_staged = _join_cond_pe_staged_jaxpr_inputs(
+      branches_staged, all_res_avals, res_avals_per_branch)
+  assert all(all(map(core.typematch, j.out_avals, branches_known[0].out_avals))
+             for j in branches_known[1:])
+
+  # Instantiate all inputs (b/c jaxpr_staged takes all inputs, corresponding to
+  # passing in_inst argument to partial_eval_jaxpr_custom above).
+  new_inst = [x for x, inst in zip(eqn.invars, inst_in)
+              if type(x) is core.Var and not inst]
+  inst_in = [True] * len(inst_in)
+
+  # Create residual variables.
+  newvar = core.gensym()
+  res_binders = map(newvar, all_res_avals)
+
+  # Build the known eqn.
+  ins_known, _ = partition_list(unks_in, eqn.invars)  # includes index invar
+  out_binders_known, _ = partition_list(unks_out, eqn.outvars)
+  linear_known = [l for l, uk in zip(eqn.params['linear'], ops_uk) if not uk]
+  params_known = dict(branches=branches_known, linear=tuple(linear_known))
+  eqn_known = pe.new_jaxpr_eqn(
+      ins_known, [*out_binders_known, *res_binders], cond_p, params_known,
+      core.no_effects, eqn.source_info)
```

Comment:
Should join the branch effects here and below, right?

Reply:
Oops, thanks! Forgot to update it to reflect #11641. Hoisted by my own petard.